# Project 3:  Exploring eBay Car Sales Data

In [2]:
import numpy as np
import pandas as pd
import datetime as dt

In [3]:
autos = pd.read_csv('Downloads/autos.csv', encoding = 'latin1')

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

- total 20 columns
- null value exists in 5 column: 6, 8, 10, 13,15
- column 15 with the most null value, non-null ~80,6%
- column name in camelcase

# Rename columns

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pictures', 'postal_code',
       'last_seen']

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerps   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Initial Exploration and Cleaning

In [11]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox       powerps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

In [12]:
autos['nr_pictures'].value_counts()

nr_pictures
0    50000
Name: count, dtype: int64

In [13]:
autos['seller'].value_counts()

seller
privat        49999
gewerblich        1
Name: count, dtype: int64

In [14]:
autos['offer_type'].value_counts()

offer_type
Angebot    49999
Gesuch         1
Name: count, dtype: int64

In [15]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [16]:
autos['odometer'].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

- nr_pictures only have 0 value --> can be dropped
- price, odometer can be converted to numberic

In [18]:
autos['price'] = autos['price'].str.replace('$','')
autos['price'] = autos['price'].str.replace(',','').astype(float)

In [19]:
autos['odometer'] = autos['odometer'].str.replace('km','')
autos['odometer'] = autos['odometer'].str.replace(',','').astype(float)

In [20]:
autos.rename(columns = {'odometer':'odometer_km'}, inplace = True)

In [21]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [22]:
autos['price'].value_counts().sort_index(ascending = True).head()
#autos['price'].value_counts().sort_index(ascending = False).head()

price
0.0    1421
1.0     156
2.0       3
3.0       1
5.0       2
Name: count, dtype: int64

- Given that eBay is an auction site, there could legitimately be items where the opening bid is 1 items, but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.
--> consider items with price over 350k are outliers

In [24]:
autos = autos[autos['price'].between(1,351000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [25]:
autos['odometer_km'].describe()

count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [26]:
autos['odometer_km'].value_counts().sort_index(ascending = True)

odometer_km
5000.0        836
10000.0       253
20000.0       762
30000.0       780
40000.0       815
50000.0      1012
60000.0      1155
70000.0      1217
80000.0      1415
90000.0      1734
100000.0     2115
125000.0     5057
150000.0    31414
Name: count, dtype: int64

odometer should have been rounded

## Exploring the Date Columns

In [29]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: proportion, dtype: float64

data was crawled daily from 05-03-2016 to 07-04-2016

In [31]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: proportion, Length: 76, dtype: float64

oldest created date is 11-06-2015, most of the ads created lately

In [33]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

last_seen
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: proportion, dtype: float64

the last three days have biggest proportion

In [35]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

registration year of 1000 and 9999 seem unrealistic

## Dealing with Incorrect Registration Year Data

In [38]:
autos = autos[autos['registration_year'].between(1900,2016)]

In [39]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

registration_year
1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028063
2013    0.017202
2014    0.014203
2015    0.008397
2016    0.026135
Name: proportion, Length: 78, dtype: float64

# Exploring Price by Brand

In [41]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [42]:
autos["brand"].value_counts(normalize=True)

brand
volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.00

Some of the most popular branch listed: volkswagen, bmw, opel, mercedes_benz,...

In [44]:
brand_count = autos["brand"].value_counts(normalize=True)
selected_brand = brand_count[brand_count> 0.05].index
print(selected_brand)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object', name='brand')


In [45]:
brand_dict = {}
for brand in selected_brand:
    brand_df = autos[autos['brand'] == brand]
    mean = brand_df['price'].mean()
    brand_dict[brand] = mean
print(brand_dict)

{'volkswagen': 5402.410261610221, 'bmw': 8332.820517811953, 'opel': 2975.2419354838707, 'mercedes_benz': 8628.450366422385, 'audi': 9336.687453600594, 'ford': 3749.4695065890287}


bmw, mercedes, audi are more expensive than volkswagen, opel, ford

# Storing Aggregate Data in a DataFrame

In [48]:
mileage_dict = {}
for brand in selected_brand:
    brand_df = autos[autos['brand'] == brand]
    mean = brand_df['odometer_km'].mean()
    mileage_dict[brand] = mean
print(mileage_dict)

{'volkswagen': 128707.15879132022, 'bmw': 132572.51313996495, 'opel': 129310.0358422939, 'mercedes_benz': 130788.36331334666, 'audi': 129157.38678544914, 'ford': 124266.01287159056}


In [49]:
mileage_sr = pd.Series(mileage_dict)
price_sr = pd.Series(brand_dict)
new_df = pd.DataFrame(mileage_sr, columns = ['mean_mileage'])
new_df['mean_price'] = price_sr
new_df

mean_mileage   mean_price
volkswagen     128707.158791  5402.410262
bmw            132572.513140  8332.820518
opel           129310.035842  2975.241935
mercedes_benz  130788.363313  8628.450366
audi           129157.386785  9336.687454
ford           124266.012872  3749.469507

mean mileage and mean price do not have much colleration

# Further cleaning

Identify categorical data that uses german words, translate them and map the values to their english counterparts

In [53]:
autos['unrepaired_damage'].value_counts()

unrepaired_damage
nein    33834
ja       4540
Name: count, dtype: int64

In [54]:
autos.loc[autos['unrepaired_damage'] == 'nein',['unrepaired_damage']] = 'no'
autos.loc[autos['unrepaired_damage'] == 'ja',['unrepaired_damage']] = 'yes'
autos['unrepaired_damage'].value_counts()

unrepaired_damage
no     33834
yes     4540
Name: count, dtype: int64

In [55]:
autos['gearbox'].value_counts()

gearbox
manuell      34715
automatik     9856
Name: count, dtype: int64

In [56]:
autos.loc[autos['gearbox'] == 'manuell',['gearbox']] = 'manual'
autos.loc[autos['gearbox'] == 'automatik',['gearbox']] = 'automatic'
autos['gearbox'].value_counts()

gearbox
manual       34715
automatic     9856
Name: count, dtype: int64

Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321

In [58]:
autos['date_crawled'] = (pd.to_datetime(autos['date_crawled']).dt.strftime('%Y%m%d')).astype(int)
autos.head()

date_crawled                                               name  seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2      20160326                         Volkswagen_Golf_1.6_United  privat   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  5000.0  control          bus               2004     manual   
1    Angebot  8500.0  control    limousine               1997  automatic   
2    Angebot  8990.0     test    limousine               2009     manual   
3    Angebot  4350.0  control   kleinwagen               2007  automatic   
4    Angebot  1350.0     test        kombi               2003     manual   

   powerps   model  odometer_km  registration_month fuel_type       brand  \
0      158  andere     150000.0                   3       lpg     peugeot   
1      286     7er     150000.0                   6    benzin         bmw   
2      102    golf      70000.0                   7    benzin  volkswagen   
3       71  fortwo      70000.0                   6    benzin       smart   
4        0   focus     150000.0                   7    benzin        ford   

  unrepaired_damage           ad_created  nr_pictures  postal_code  \
0                no  2016-03-26 00:00:00            0        79588   
1                no  2016-04-04 00:00:00            0        71034   
2                no  2016-03-26 00:00:00            0        35394   
3                no  2016-03-12 00:00:00            0        33729   
4                no  2016-04-01 00:00:00            0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

See if there are particular keywords in the name column that you can extract as new columns

In [60]:
split_name = autos['name'].str.split('_',n=2,expand=True)
split_name

0       1                                                  2
0         Peugeot     807                               160_NAVTECH_ON_BOARD
1             BMW    740i                  4_4_Liter_HAMANN_UMBAU_Mega_Optik
2      Volkswagen    Golf                                         1.6_United
3           Smart   smart            fortwo_coupe_softouch/F1/Klima/Panorama
4            Ford   Focus  1_6_Benzin_TÜV_neu_ist_sehr_gepflegt.mit_Klima...
...           ...     ...                                                ...
49995        Audi      Q5           3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon
49996        Opel   Astra   F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!
49997        Fiat     500                              C_1.2_Dualogic_Lounge
49998        Audi      A3                         2.0_TDI_Sportback_Ambition
49999        Opel  Vectra                                            1.6_16V

[46681 rows x 3 columns]

In [61]:
autos['new_name'] = split_name[0] + '_' + split_name[1]

In [62]:
autos['new_name'].value_counts().head(20)

new_name
Mercedes_Benz        3159
Volkswagen_Golf      1821
Opel_Corsa           1195
Opel_Astra           1072
Audi_A4              1006
Volkswagen_Passat     730
Volkswagen_Polo       688
Audi_A6               665
Ford_Focus            652
Audi_A3               646
Ford_Fiesta           585
VW_Golf               468
Renault_Twingo        436
Peugeot_206           402
Ford_Mondeo           381
Opel_Vectra           379
Renault_Clio          362
VW_Polo               335
Opel_Zafira           302
Fiat_Punto            271
Name: count, dtype: int64

# Further analysis

Find the most common brand/model combinations

In [65]:
autos['brand_model'] = autos['brand'] + '_' + autos['model']

In [66]:
autos['brand_model'].value_counts().head(20)

brand_model
volkswagen_golf           3707
bmw_3er                   2615
volkswagen_polo           1609
opel_corsa                1592
volkswagen_passat         1349
opel_astra                1348
audi_a4                   1231
mercedes_benz_c_klasse    1136
bmw_5er                   1132
mercedes_benz_e_klasse     958
audi_a3                    825
audi_a6                    797
ford_focus                 762
ford_fiesta                722
volkswagen_transporter     674
renault_twingo             615
peugeot_2_reihe            600
smart_fortwo               550
opel_vectra                544
mercedes_benz_a_klasse     539
Name: count, dtype: int64

Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.

In [68]:
autos['odometer_km'].value_counts()

odometer_km
150000.0    30085
125000.0     4857
100000.0     2058
90000.0      1673
80000.0      1375
70000.0      1187
60000.0      1128
50000.0       993
40000.0       797
5000.0        785
30000.0       760
20000.0       742
10000.0       241
Name: count, dtype: int64

In [69]:
odo_list = autos['odometer_km'].unique()

In [70]:
odo_dict = {}
for i in odo_list:
    p = autos[autos['odometer_km'] == i]['price'].mean()
    odo_dict[i] = p
odo_sr = pd.Series(odo_dict)
odo_df = pd.DataFrame(odo_sr,columns = ['mean_price'])
odo_df.sort_index()    

mean_price
5000.0     8873.515924
10000.0   20550.867220
20000.0   18448.477089
30000.0   16608.836842
40000.0   15499.568381
50000.0   13812.173212
60000.0   12385.004433
70000.0   10927.182814
80000.0    9721.947636
90000.0    8465.025105
100000.0   8132.697279
125000.0   6214.022030
150000.0   3767.927107

--> high mileage seems to be cheaper than, except for very low mileage of 5000 km

How much cheaper are cars with damage than their non-damaged counterparts

In [73]:
autos[autos['unrepaired_damage'] == 'yes']['price'].mean()

2241.146035242291

In [74]:
autos[autos['unrepaired_damage'] == 'no']['price'].mean()

7164.033102796004